moby/moby
# Adjust StdCopy so that it can handle headless streams

**- What I did**

When we process container logs, most people will follow the example given in the official tutorial (for example: https://docs.docker.com/engine/api/sdk/examples/). This is effective in most cases. It uses the StdCopy function to distribute logs to dstout and dsterr, which seems to be a very cool feature.
However, when running the container with the -tty flag, StdCopy will not work properly. It will report "Unrecognized input header: xxx" (strictly speaking, this is not a bug). This is a very big doubt for beginners.

When we read the docker source code, we found that when tty:true, the source code is as follows: "outStream := io.Writer(wf) errStream := outStream", this is a stream without a head. When tty: false,The source code is as follows: "sysErrStream=stdcopy.NewStdWriter(outStream,stdcopy.Systemerr)
errStream=stdcopy.NewStdWriter(outStream, stdcopy.Stderr)", this is a stream with a header. After reading the source code, we can understand when to use StdCopy and when to use io.Copy. However, not everyone will read the source code. More people just call the API to complete their work.
Personally think that docker can provide a more compatible StdCopy for everyone to use, which is more friendly to beginners.
Therefore, I modified the StdCopy function, hoping that it can be compatible with the handling of container logs with tty flags turned on.

**- How I did it**

When parsing the head of the stream, determine whether StdType belongs to one of Stdin, Stdout, Stderr, and Systemerr. If it is one of them, shunt them. Otherwise, think of it as a stream without a header, copy the content directly to ‘dstdout’

**- How to verify it**
`

	package stdcopy // import "github.com/docker/docker/pkg/stdcopy"
	
	import (
		"bytes"
		"errors"
		"io"
		"io/ioutil"
		"strings"
		"testing"
	)

	func TestStdCopyWriteAndRead(t *testing.T) {
		stdOutBytes := []byte(strings.Repeat("o", startingBufLen))
		stdErrBytes := []byte(strings.Repeat("e", startingBufLen))
		buffer, err := getSrcBuffer(stdOutBytes, stdErrBytes)
		if err != nil {
			t.Fatal(err)
		}
		written, err := StdCopy(ioutil.Discard, ioutil.Discard, buffer)
		if err != nil {
			t.Fatal(err)
		}
		expectedTotalWritten := len(stdOutBytes) + len(stdErrBytes)
		if written != int64(expectedTotalWritten) {
			t.Fatalf("Expected to have total of %d bytes written, got %d", expectedTotalWritten, written)
		}
	}
	
	func TestStdCopyWriteAndReadTtyStream(t *testing.T) {
		stdOutBytes := []byte(strings.Repeat("o", startingBufLen))
		srcInput := new(bytes.Buffer)
	
		writtenum, err := srcInput.Write(stdOutBytes)
	
		t.Logf("write %d bytes",writtenum)
		if err != nil {
			t.Fatal(err)
		}
	
		written, err := StdCopy(ioutil.Discard, ioutil.Discard, srcInput)
		if err != nil {
			t.Fatal(err)
		}
		t.Skipf("read %d bytes",written)
	}
`

The above code is also written in the stdcopy_test.go file.

**- Description for the changelog**

Enables StdCopy to handle both headed and headless streams

**- A picture of a cute animal (not mandatory but encouraged)**
![lovelybee](https://user-images.githubusercontent.com/11941914/84561515-b2c61800-ad7f-11ea-8550-1081a4850a35.jpg)
